File: swift/beer-song/Sources/BeerSong/BeerSong.swift
struct BeerSong {
    let numberOfBeerBottles: Int
    
    func generateVersesOfBeerSong() -> String {
        var count = numberOfBeerBottles-1
        var returnString = getVerse(for: numberOfBeerBottles)
        while count >= 0 {
            
            returnString += "\n\n\(getVerse(for: count))"

            count -= 1
        }
        return returnString
    }
    
    private func getVerse(for num: Int) -> String {
        let firstVerse: String
        var secondVerse: String
        
        if num > 0 {
            firstVerse = "\(num) \(num > 1 ? "bottles" : "bottle") of beer on the wall, \(num) \(num > 1 ? "bottles" : "bottle") of beer."
        } else {
            firstVerse = "No more bottles of beer on the wall, no more bottles of beer."
        }
                
        if num > 0 {
            secondVerse = "Take one down and pass it around, "
        } else {
            secondVerse = "Go to the store and buy some more, "
        }
        
        let remaining: Int = num-1 >= 0 ? num-1 : 99
        
        let secondHalf: String
        switch remaining {
        case 0:
            secondHalf = "no more bottles"
        default:
            secondHalf = "\(remaining) \(remaining > 1 ? "bottles" : "bottle")"
        }
        
        secondVerse += "\(secondHalf) of beer on the wall."

        return "\(firstVerse)\n\(secondVerse)"
    }
}

File: kotlin/beer-song/src/main/kotlin/BeerSong.kt
object BeerSong {
    fun verses(startBottles: Int, takeDown: Int): String {
        var result = ""
        var start = startBottles
        
        val times = takeDown - start
        for (i in 0 until times+1) {
            val firstVerse = if (start > 1) {
                "$start bottles of beer on the wall, $start bottles of beer.\n"
            } else if (start == 1) {
                "$start bottle of beer on the wall, $start bottle of beer.\n"
            } else {
                "No more bottles of beer on the wall, no more bottles of beer.\n"
            }
            val secondVerse: String
            val remain = start - 1

            if (remain > 1) {
                secondVerse = "Take one down and pass it around, $remain bottles of beer on the wall.\n"
            } else if (remain == 1) {
                secondVerse = "Take it down and pass it around, no more bottles of beer on the wall.\n"
            } else {
                secondVerse = "Go to the store and buy some more, 99 bottles of beer on the wall.\n"
            }
            
            result = result + firstVerse + secondVerse
            start = remain
        }
        
        return result
    }
}
